### ISSUE TITLE
Use STGCN++ to train on own 3d dataset, but the top1_acc always is 0.1.

### ISSUE THREAD

**Issue by chz8bit** [kennymckormick/pyskl#137]
I get a video dataset by myself.  Then I get 2d joint-keypoints by `tools/data/custom_2d_skeleton.py` 
After that, I did the 2d-to-3d PHE to get 3d joint-keypoints. The generated format is Human3.6M. 
I mapped the 17 joint-keypoints of Human3.6M to the 3d joint-keypoints of NTURGB+D (the points of the hand and foot coincide). 

Lastly, I used STGCN++to train, but the top1 has been maintained at 0.1. 
What is the matter?

My guess is that the data obtained by 2d-to-3d PHE is in spatial correspondence with NTURGB+D. 

I annotated `if not ((np.allclose(j1, 0) or np.allclose(j2, 0)) and link[0] ! = 1 and link[1] ! = 1):`  for overlapping points .  And I see `PreNormalize3D` is used to Pre-Normalization for NTURGB+D, but it's useless for my 3d joint-keypoints dataset. So I did not use it.

![image](https://user-images.githubusercontent.com/117361008/220122410-3a7e1471-a234-42f8-a0d6-2780d2561fd9.png)

**Comment by chz8bit** [comment]
This is my config.
```
model = dict(
    type='RecognizerGCN',
    backbone=dict(
        type='STGCN',
        gcn_adaptive='init',
        gcn_with_res=True,
        tcn_type='mstcn',
        graph_cfg=dict(layout='nturgb+d', mode='spatial')),
    cls_head=dict(type='GCNHead', num_classes=4, in_channels=256))

dataset_type = 'PoseDataset'
ann_file = 'dataset_cut/teacherDataset_p3d1.pkl'
train_pipeline = [
    dict(type='GenSkeFeat', dataset='nturgb+d', feats=['j']),
    dict(type='UniformSample', clip_len=10),
    dict(type='PoseDecode'),
    dict(type='FormatGCNInput', num_person=1),
    dict(type='Collect', keys=['keypoint', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['keypoint'])
]
val_pipeline = [
    dict(type='GenSkeFeat', dataset='nturgb+d', feats=['j']),
    dict(type='UniformSample', clip_len=10, num_clips=1),
    dict(type='PoseDecode'),
    dict(type='FormatGCNInput', num_person=1),
    dict(type='Collect', keys=['keypoint', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['keypoint'])
]
test_pipeline = [
    dict(type='GenSkeFeat', dataset='nturgb+d', feats=['j']),
    dict(type='UniformSample', clip_len=10, num_clips=20),
    dict(type='PoseDecode'),
    dict(type='FormatGCNInput', num_person=1),
    dict(type='Collect', keys=['keypoint', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['keypoint'])

# optimizer
optimizer = dict(type='SGD', lr=0.1, momentum=0.9, weight_decay=0.0005, nesterov=True)
optimizer_config = dict(grad_clip=None)
# learning policy
lr_config = dict(policy='CosineAnnealing', min_lr=0, by_epoch=False)
total_epochs = 2000
checkpoint_config = dict(interval=50)
evaluation = dict(interval=1, metrics=['top_k_accuracy'])
log_config = dict(interval=100, hooks=[dict(type='TextLoggerHook')])
```

**Comment by kennymckormick** [comment]
Hi, @chz8bit , what is the size of your training set?

**Comment by chz8bit** [comment]
> Hi, @chz8bit , what is the size of your training set?

Hi, thanks for your answer!

My dataset contains 95 video clips with a total of 13,000 frames. I guessed the dataset is too small, so I tried to train based on the pre-trained model, [stgcn++/stgcn++_ntu60_xsub_3dkp/j/j.py](http://download.openmmlab.com/mmaction/pyskl/ckpt/stgcnpp/stgcnpp_ntu60_xsub_3dkp/j.pth), in the last few hours. But the `top1_acc` is the same as before, still 0.1. 

This is my config.
```
model = dict(
    type='RecognizerGCN',
    backbone=dict(
        type='STGCN',
        gcn_adaptive='init',
        gcn_with_res=True,
        tcn_type='mstcn',
        graph_cfg=dict(layout='nturgb+d', mode='spatial')),
    cls_head=dict(type='GCNHead', num_classes=4, in_channels=256))

dataset_type = 'PoseDataset'
ann_file = 'dataset_cut/teacherDataset_p3d1.pkl'
load_from = 'stgcnplus/j.pth'

train_pipeline = [
    dict(type='PreNormalize3D'),
    dict(type='GenSkeFeat', dataset='nturgb+d', feats=['j']),
    dict(type='UniformSample', clip_len=20),
    dict(type='PoseDecode'),
    dict(type='FormatGCNInput', num_person=1),
    dict(type='Collect', keys=['keypoint', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['keypoint'])
]
val_pipeline = [
    dict(type='PreNormalize3D'),
    dict(type='GenSkeFeat', dataset='nturgb+d', feats=['j']),
    dict(type='UniformSample', clip_len=20, num_clips=1),
    dict(type='PoseDecode'),
    dict(type='FormatGCNInput', num_person=1),
    dict(type='Collect', keys=['keypoint', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['keypoint'])
]
test_pipeline = [
    dict(type='PreNormalize3D'),
    dict(type='GenSkeFeat', dataset='nturgb+d', feats=['j']),
    dict(type='UniformSample', clip_len=20, num_clips=10),
    dict(type='PoseDecode'),
    dict(type='FormatGCNInput', num_person=1),
    dict(type='Collect', keys=['keypoint', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['keypoint'])
]
data = dict(
    videos_per_gpu=16,
    workers_per_gpu=2,
    test_dataloader=dict(videos_per_gpu=1),
    train=dict(
        type='RepeatDataset',
        times=10,
        dataset=dict(type=dataset_type, ann_file=ann_file, pipeline=train_pipeline, split='train')),
    val=dict(type=dataset_type, ann_file=ann_file, pipeline=val_pipeline, split='test'),
    test=dict(type=dataset_type, ann_file=ann_file, pipeline=test_pipeline, split='test'))

# optimizer
optimizer = dict(type='SGD', lr=0.001, momentum=0.9, weight_decay=0.0005, nesterov=True)
optimizer_config = dict(grad_clip=None)
# learning policy
lr_config = dict(policy='CosineAnnealing', min_lr=0, by_epoch=False)
total_epochs = 500
checkpoint_config = dict(interval=50)
evaluation = dict(interval=10, metrics=['top_k_accuracy'])
log_config = dict(interval=100, hooks=[dict(type='TextLoggerHook')])

# runtime settings
log_level = 'INFO'
work_dir = './work_dirs/teacher/j3d_2-21_prePreNormalize3D'
```

This is the log.

```
{"mode": "val", "epoch": 10, "iter": 1, "lr": 0.001, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 20, "iter": 1, "lr": 0.001, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 30, "iter": 1, "lr": 0.00099, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 40, "iter": 1, "lr": 0.00098, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 50, "iter": 1, "lr": 0.00098, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 60, "iter": 1, "lr": 0.00096, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 70, "iter": 1, "lr": 0.00095, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 80, "iter": 1, "lr": 0.00094, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 90, "iter": 1, "lr": 0.00092, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 100, "iter": 1, "lr": 0.0009, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 110, "iter": 1, "lr": 0.00089, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 120, "iter": 1, "lr": 0.00086, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 130, "iter": 1, "lr": 0.00084, "top1_acc": 0.1, "top5_acc": 1.0}
{"mode": "val", "epoch": 140, "iter": 1, "lr": 0.00082, "top1_acc": 0.1, "top5_acc": 1.0}
```

**Comment by chz8bit** [comment]
If you need this dataset for testing, I am willing to provide it by email.

**Comment by kennymckormick** [comment]
I'm willing to have a closer look if you sent it to my email.

**Comment by chz8bit** [comment]
I have sent the dataset to your email, which is the gmail email on your homepage.

I hope I can get your help!

I have a guess, `frame_dir` will affect the training. But I see it supports cross-subject via some paper.  I have no idea.   :(

**Comment by chz8bit** [comment]
I found an interesting thing. I put my 2d dataset into poseC3D. 
`configs/posec3d/c3d_light_ntu60_xsub/joint.py` works, but `configs/posec3d/c3d_light_ntu60_xsub/limb.py ` does not work. The result of `poseC3D_limb` is the same as before, `top1_acc: 0.1000`. 

I guess the limb is the cause of these results. 


This is a part of joint.json

```
{"mode": "val", "epoch": 320, "iter": 1, "lr": 0.11501, "top1_acc": 0.7, "top5_acc": 1.0, "mean_class_accuracy": 0.89286}
{"mode": "val", "epoch": 330, "iter": 1, "lr": 0.10381, "top1_acc": 0.8, "top5_acc": 1.0, "mean_class_accuracy": 0.92857}
{"mode": "val", "epoch": 340, "iter": 1, "lr": 0.09299, "top1_acc": 0.9, "top5_acc": 1.0, "mean_class_accuracy": 0.96429}
{"mode": "val", "epoch": 350, "iter": 1, "lr": 0.08259, "top1_acc": 0.7, "top5_acc": 1.0, "mean_class_accuracy": 0.89286}
{"mode": "val", "epoch": 360, "iter": 1, "lr": 0.07265, "top1_acc": 0.8, "top5_acc": 1.0, "mean_class_accuracy": 0.92857}
{"mode": "val", "epoch": 370, "iter": 1, "lr": 0.06322, "top1_acc": 0.7, "top5_acc": 1.0, "mean_class_accuracy": 0.89286}
{"mode": "val", "epoch": 380, "iter": 1, "lr": 0.05433, "top1_acc": 0.7, "top5_acc": 1.0, "mean_class_accuracy": 0.89286}
```


This is a part of limb.json

```
{"mode": "val", "epoch": 96, "iter": 1, "lr": 0.27162, "top1_acc": 0.1, "top5_acc": 1.0, "mean_class_accuracy": 0.25}
{"mode": "val", "epoch": 97, "iter": 1, "lr": 0.26927, "top1_acc": 0.1, "top5_acc": 1.0, "mean_class_accuracy": 0.25}
{"mode": "val", "epoch": 98, "iter": 1, "lr": 0.2669, "top1_acc": 0.1, "top5_acc": 1.0, "mean_class_accuracy": 0.25}
{"mode": "val", "epoch": 99, "iter": 1, "lr": 0.26453, "top1_acc": 0.1, "top5_acc": 1.0, "mean_class_accuracy": 0.25}
{"mode": "val", "epoch": 100, "iter": 1, "lr": 0.26214, "top1_acc": 0.1, "top5_acc": 1.0, "mean_class_accuracy": 0.25}
{"mode": "val", "epoch": 101, "iter": 1, "lr": 0.25975, "top1_acc": 0.1, "top5_acc": 1.0, "mean_class_accuracy": 0.25}
{"mode": "val", "epoch": 102, "iter": 1, "lr": 0.25735, "top1_acc": 0.1, "top5_acc": 1.0, "mean_class_accuracy": 0.25}
{"mode": "val", "epoch": 103, "iter": 1, "lr": 0.25494, "top1_acc": 0.1, "top5_acc": 1.0, "mean_class_accuracy": 0.25}
```

**Comment by kennymckormick** [comment]
'frame_dir' will not affect training, I have read your data but still not sure about the problem, some reasonable guess can be: 
1. You only have a small number of training samples.
2. Your keypoint data looks weird: the magnitude of the 3D keypoint is ~10^2. I'm not familiar with such kinds of data.

**Comment by chz8bit** [comment]
> 'frame_dir' will not affect training, I have read your data but still not sure about the problem, some reasonable guess can be:
> 
> 1. You only have a small number of training samples.
> 2. Your keypoint data looks weird: the magnitude of the 3D keypoint is ~10^2. I'm not familiar with such kinds of data.

Thanks for your help!

1. Yes, I only have a small number of training samples. Maybe I should get larger dataset.   :  )
2. ` the magnitude of the 3D keypoint is ~10^2`. The size obtained from the 2d-to-3d PHE experiment.  Perhaps I should normalize it. I checked the NTURGB data, where x and y are between -1 and 1, and z is between 3.0 and 4.0. I think it's because of the sensor. I am using the HPE code base on Human3.6, and the sensors used in both are different. 

Thanks for your help! :>
I will try to get a larger dataset and read the code about normalization.

**Comment by leviethung2103** [comment]
@chz8bit  Have you resolve this problem ?
